### ISSUE TITLE
Search this area of the map button: focus management and screenreaders

### ISSUE THREAD

**Issue by mmiddaugh** [department-of-veterans-affairs/va.gov-team#21256]
## Issue Description
#21958 addressed the connection between the appearance of the "Search this area of the map" button and interaction with the map (along with the refactoring).
#22960 addresses state change screenreader announcements 

**This issue will address the addition of instructions to support all users**

## Tasks
- [ ] Add information about the button to the zoom control instructions: 
       > ~_This is an interactive map. Zoom in or out using the plus and minus buttons. Use your keyboard to navigate up, down, left, and right in the map. A search this area of map button will become available when you are zoomed in to an area of 500 miles or less.~ _updated to include mobile instructions_: "Zoom in or out using the plus and minus buttons or by pinching and pulling with your fingers. Use your keyboard or drag your fingers to navigate up, down, left, and right in the map. You can search areas of up to 500 miles or less on the map."
We'll need to provide a label for the map using aria-label and instructions using aria-describedby.
- [ ] Add a descriptive aria-label that makes it clear that users can interact with the map
- [ ] Add visible instructions 

## Acceptance Criteria 
- [ ] Upon tabbing into the map, instructions are provided to Veterans using screen readers. 
- [ ] Instructions are accessible to sighted users with cognitive disabilities who may need explicit guidance on how to use it

**Comment by mmiddaugh** [comment]
Validate [Focus should go to results subhead following page refresh (FL)](https://github.com/department-of-veterans-affairs/va.gov-team/issues/21262) following this work.

**Comment by mmiddaugh** [comment]
@joshkimux will you please confirm that announcing/explaining the "Search this area of the map" button the first time the map receives focus will suffice, rather than each time it appears/disappears? Does the suggested text in the description cover the necessary information? (It may need content polishing) -> _something like...._This is an interactive map. You may zoom in/out of the map using +/- . You may search an area of 500 miles or less by tabbing to the "Search this area of the map" button._

**Comment by joshkimux** [comment]
> Search this area of the map button will not display unless the Veteran has interacted with the map by panning or zooming.

### I think there's 3 ways of solving for this, which I'll document probable requirements for below. ###

#### _Always_ show the "Search this area of the map button":
- We would need to provide additional instructions that results can only be provided at an area of 500m or less to sighted users (which would then also cover screen reader users). e.g. A small text description above the map _This is an interactive map. Zoom in or out using the plus and minus buttons. Use your keyboard to navigate up, down, left, and right in the map. Use the search this area of map button to refresh results._ 
- We would also need to provide an error message to provide feedback that the button will not display results unless the 500m or less condition is met.

#### Show and hide the "Search this area of the map button" (current state): 
- We would need to provide screen reader only instructions upon entering the map to explain how to interact with it. For example, _This is an interactive map. Zoom in or out using the plus and minus buttons. Use your keyboard to navigate up, down, left, and right in the map. A search this area of map button will become available after the zoom in buttons when you are zoomed [in by level 5] **OR** [in to an area of 500 miles or less]_
- We would need to announce zoom levels or miles upon usage of the plus and minus buttons. I personally hypothesize that zoom levels will be easier to remember compared to miles, but I can also see how miles may be more understandable to users. This would depend on what we know from research @lildziner 

#### _Disable_ the ability to zoom out beyond 500 miles and _always_ show the "Search this area of the map button"
- We would need to provide screen reader only instructions upon entering the map to explain how to interact with it. For example, _This is an interactive map. Zoom in or out using the plus and minus buttons. Use your keyboard to navigate up, down, left, and right in the map. Use the search this area of map button to refresh results._

From a UX perspective, I personally recommend the last option as I think it's the least verbose and most straightforward. Would love @mmiddaugh 's thoughts and @mpelzsherman 's technical feedback as y'all are the experts 🙇 

----

> Search this area of the map button receives focus when displayed. (Confirm with Josh)

After testing, I think it's best that the search this area of the map _does not_ receive focus when displayed as it may unpredictably jump keyboard users out to the map after conducting a regular search (they may prefer to tab down to the list first)

**Comment by lildziner** [comment]
I like the 3rd option:•the focus goes back to the search results instead of the "Search this area of the map button". Having the search button ever present breaks the popular map conventions that I've interacted with out in the wild by not appearing/disappearing but I think it does simplify the interactivity. Also, we allow users to hit search in above as many times as they want even if they haven't edited their search parameters so it will be consistent for both searches

**Comment by mmiddaugh** [comment]
@joshkimux 
We've implemented option 3 as described above - using a disabled state and text change when the search area becomes too large. As we add the screen reader/focus management, we have the following questions: 
1. When, if ever, should the button automatically get focus?
2. If the button returns to an active state and is announced, do instructions need to be given on how to access it?
3. Are the instructions regarding map interactions acceptable as written in the description?

**Comment by joshkimux** [comment]
@mmiddaugh 

1. The button should never automatically get focus

As for 2 and 3... this is tricky. I started by testing what the interaction would be like with instructions upon tabbing into the map using an `aria-label` and invisible `aria-describedby`

> Find VA locations using an interactive map. 
> Zoom in or out using the plus and minus buttons. Use your keyboard to navigate up, down, left, and right in the map. A search this area of map button will become available when you are zoomed in to an area of 500 miles or less.

But I have the following concerns/questions post-testing:

- Would we be able to provide a distance by radius? If we provide instructions that the search this area of map will only be available at an area of 500 miles or less, it would make sense to take Angela's feedback in providing an actual radius.

- Would these instructions be valuable to everyone using this? Knowing we have an older audience and Veterans with cognitive and learning disabilities, it may be helpful to provide explicit instructions to everyone. This description could be used as a visible `aria-describedby` to the map. For example:

<img width="804" alt="Screen Shot 2021-04-07 at 2 33 42 PM" src="https://user-images.githubusercontent.com/14154792/113916872-81a35200-97ae-11eb-8853-59db29c0fd3b.png">

- It would be more accessible and human-centered to always keep the button active. Doing so would prevent the scenario where a screen reader is lost trying to find a button that isn't on the page yet even after having listened to the instructions. It also lets us provide explicit error messaging on how to fix the issue. For mobile screen reader users, it would provide a consistent touch location.

- On the note of _mobile screen reader_ usage, I made the mistake of writing instructions from a desktop centric perspective. Instead of just providing keyboard instructions, we would need to provide mobile screen reader instructions too (I think it may be the more popular use case). For example, 

> "Zoom in or out using the plus and minus buttons or by pinching and pulling with your fingers. Use your keyboard or drag your fingers to navigate up, down, left, and right in the map. You can search areas of up to 500 miles or less on the map."

**Comment by joshkimux** [comment]
@mmiddaugh To answer more short term questions:

2. If the button returns to an active state and is announced, do instructions need to be given on how to access it?
No. As long as instructions on the expected behavior is provided upon entering the map, further instructions shouldn't be needed every time it appears and disappears. Post-testing, I think in the future it would be best to update this to always be active along with error messaging.

3. Are the instructions regarding map interactions acceptable as written in the description?
No. I think we should provide instructions for mobile screen reader users as well. For example,

> Zoom in or out using the plus and minus buttons or by pinching and pulling with your fingers. Use your keyboard or drag your fingers to navigate up, down, left, and right in the map. You can search areas of up to 500 miles or less on the map.

@AngelaFowler82 would be the best person to reference for guidance on language there.

**Comment by mmiddaugh** [comment]
Thanks @joshkimux!

> Would we be able to provide a distance by radius?
We have this issue in the backlog for future prioritization and think this is a good step in that direction. 

Great catch on the mobile instructions! We were all thinking desktop-centric on that one. 

Leyda is beginning to formulate a plan for accessibility research specifically with users of assistive tech. We know we'll get invaluable feedback to inform future iterations.

**Comment by joshkimux** [comment]
@lildziner @mpelzsherman 

Some additional accessibility rules/context that might help for implementation.

We'll need to provide a label for the map using `aria-label` and instructions using `aria-describedby`.

## 1. Add a descriptive `aria-label` that makes it clear that users can interact with the map

```html
<map
  aria-label="Find VA locations on an interactive map" 
></map>
```

## 2. Either add _visible_ instructions (recommended) OR make them `sr-only`

### Visible
```html
<p id="instructions">
  Zoom in or out using the plus and minus buttons or by pinching and pulling with your fingers. Use your keyboard or drag your fingers to navigate up, down, left, and right in the map. You can search areas of up to 500 miles or less on the map.
</p>
<map
  aria-label="Find VA locations on an interactive map" 
></map>
```

<img width="1175" alt="Screen Shot 2021-04-08 at 12 29 02 PM" src="https://user-images.githubusercontent.com/14154792/114062878-03a28200-9866-11eb-9504-326934be17e0.png">

Ideally, instructions would live above the map to address @mpelzsherman 's point on the page's fold. This will most likely require design iterations from @lildziner to figure out a non-intrusive way of providing instructions that don't increase the noise of the page.

### Hidden
Hidden instructions could potentially fail to meet WCAG SC 3.3.2 Labels or Instructions depending on how intuitive or unintuitive the behavior of the map is to sighted users. However, this may be a less design-intrusive and temporary solution until we figure out a better way to address visible instructions.

```html
<p class="sr-only" id="instructions">
  Zoom in or out using the plus and minus buttons or by pinching and pulling with your fingers. Use your keyboard or drag your fingers to navigate up, down, left, and right in the map. You can search areas of up to 500 miles or less on the map.
</p>
<map
  aria-label="Find VA locations on an interactive map" 
></map>
```

<img width="1085" alt="Screen Shot 2021-04-08 at 12 32 24 PM" src="https://user-images.githubusercontent.com/14154792/114063338-80356080-9866-11eb-8538-3303a727e114.png">

Instructions will still be read out to screen reader users, but won't be accessible to sighted users with cognitive disabilities who may need explicit guidance on how to use it.

## 3. Point to the instructions using `aria-describedby`

```html
<p class="sr-only" id="instructions">
  Zoom in or out using the plus and minus buttons or by pinching and pulling with your fingers. Use your keyboard or drag your fingers to navigate up, down, left, and right in the map. You can search areas of up to 500 miles or less on the map.
</p>
<map
  aria-label="Find VA locations on an interactive map" 
  aria-describedby="instructions"
></map>
```

## Example of expected behavior
Upon tabbing into the map, the `aria-label` is read out followed by the `aria-describedby`

https://user-images.githubusercontent.com/14154792/114063775-f5a13100-9866-11eb-89c4-cfdbe88c9209.mov

**Comment by joshkimux** [comment]
We've decided to stick to screen reader only instructions as outlined by the 2nd recommendation above! @lildziner @mpelzsherman

**Comment by mmiddaugh** [comment]
This issue has been discussed extensively here and in Slack. We've determined that the priority requirement here is to include screenreader instructions. Although separate mobile and desktop instructions would be ideal, the work would be complicated by device types and breakpoints.  We will proceed with a simple, global announcement for this iteration. Accessibility research is planned this quarter and can be used to inform future iterations.
